lib/api/include/irods/ticketAPI.hpp

```diff
@@ -0,0 +1,414 @@
+#ifndef IRODS_TICKET_API_HPP
+#define IRODS_TICKET_API_HPP
+
+#include <string_view>
+#include <type_traits>
+#include <iostream>
+#include <string>
+
+#undef NAMESPACE_IMPL
+#undef RxComm
+#undef rxTicketAdmin
+
+#ifdef IRODS_TICKET_ADMINISTRATION_ENABLE_SERVER_SIDE_API
+#    define NAMESPACE_IMPL server
+#    define RxComm         RsComm
+#    define rxTicketAdmin  rsTicketAdmin
+
+struct RsComm;
+#else
+#    define NAMESPACE_IMPL client
+#    define RxComm         RcComm
+#    define rxTicketAdmin  rcTicketAdmin
+
+struct RcComm;
+#endif // IRODS_TICKET_ADMINISTRATION_ENABLE_SERVER_SIDE_API
+
+/**
+ * @brief Namespace to contain the API for tickets
+ *
+ */
+namespace irods::administration::ticket
+{
+    inline struct admin_tag
+    {
+    } admin;
+
+    enum class ticket_type
+    {
+        read,
+        write
+    };
+
+    enum class ticket_property
+    {
+        uses_count,
+        write_file,
+        write_byte
+    };
+
+    void execute_ticket_operation(RxComm& conn,
+                                  std::string_view command,
+                                  std::string_view ticket_identifier,
+                                  std::string_view command_modifier1,
+                                  std::string_view command_modifier2,
+                                  std::string_view command_modifier3,
+                                  std::string_view command_modifier4,
+                                  bool run_as_admin);
+
+    /**
+     * @brief Create a ticket object
+     *
+     * @param conn Communication object
+     * @param _type Specify type for ticket
+     * @param obj_path Path for the object
+     * @param ticket_name Name of the ticket (if provided)
+     * @return int Error code
+     */
+    void create_ticket(RxComm& conn, ticket_type _type, std::string_view obj_path, std::string_view ticket_name);
+    std::string create_ticket(RxComm& conn, ticket_type _type, std::string_view obj_path);
+
+    void create_ticket(admin_tag,
+                       RxComm& conn,
+                       ticket_type _type,
+                       std::string_view obj_path,
+                       std::string_view ticket_name);
+    std::string create_ticket(admin_tag, RxComm& conn, ticket_type _type, std::string_view obj_path);
+
+    /**
+     * @brief Delete a ticket
+     *
+     * @param conn Communication Object
+     * @param ticket_name Name of the ticket
+     * or
+     * @param ticket_ID ID for the ticket that needs to be deleted
+     * @return int Error Code
+     */
+    void delete_ticket(RxComm& conn, std::string_view ticket_name);
+    void delete_ticket(RxComm& conn, int ticket_ID);
+
+    void delete_ticket(admin_tag, RxComm& conn, std::string_view ticket_name);
+    void delete_ticket(admin_tag, RxComm& conn, int ticket_ID);
+
+    struct users_constraint
+    {
+        const std::string_view user_name;
```

Comment:
Remove `const` modifier.

We can defer whether these should be const to the interfaces that use them.